crates/bevy_ui/src/stack.rs

```diff
@@ -15,102 +15,78 @@ pub struct UiStack {
-                Some(&mut new_context),
-                total_entry_count,
-            );
+    let uinodes = &mut ui_stack.uinodes;
+
+    fn update_uistack_recursively(
+        entity: Entity,
+        uinodes: &mut Vec<Entity>,
+        children_query: &Query<&Children>,
+        zindex_query: &Query<Option<&ZIndex>, (With<Node>, Without<GlobalZIndex>)>,
+    ) {
+        uinodes.push(entity);
+
+        if let Ok(children) = children_query.get(entity) {
+            let mut z_children: Vec<(Entity, i32)> = children
+                .iter()
+                .filter_map(|entity| {
+                    zindex_query
+                        .get(*entity)
+                        .ok()
+                        .map(|zindex| (*entity, zindex.map(|zindex| zindex.0).unwrap_or(0)))
+                })
+                .collect();
+            z_children.sort_by_key(|k| k.1);
+            for (child_id, _) in z_children {
+                update_uistack_recursively(child_id, uinodes, children_query, zindex_query);
+            }
         }
     }
 
-    // The node will be added either to global/parent based on its z-index type: global/local.
-    let z_index = zindex_query.get(entity).unwrap_or(&ZIndex::Local(0));
-    let (entity_context, z_index) = match z_index {
-        ZIndex::Local(value) => (parent_context.unwrap_or(global_context), *value),
-        ZIndex::Global(value) => (global_context, *value),
-    };
-
-    *total_entry_count += 1;
-    entity_context.entries.push(StackingContextEntry {
-        z_index,
-        entity,
-        stack: new_context,
-    });
-}
-
-/// Flatten `StackingContext` (z-index based UI node tree) into back-to-front entities list
-fn fill_stack_recursively(result: &mut Vec<Entity>, stack: &mut StackingContext) {
-    // Sort entries by ascending z_index, while ensuring that siblings
-    // with the same local z_index will keep their ordering. This results
-    // in `back-to-front` ordering, low z_index = back; high z_index = front.
-    stack.entries.sort_by_key(|e| e.z_index);
-
-    for entry in &mut stack.entries {
-        // Parent node renders before/behind child nodes
-        result.push(entry.entity);
-        fill_stack_recursively(result, &mut entry.stack);
+    let global_nodes = zindex_global_node_query
+        .iter()
+        .map(|(id, global_zindex, maybe_zindex)| {
+            (
+                id,
+                (
+                    global_zindex.0,
+                    maybe_zindex.map(|zindex| zindex.0).unwrap_or(0),
+                ),
+            )
+        });
+
+    let mut root_nodes: Vec<_> = root_node_query
+        .iter()
+        .map(|(root_id, maybe_global_zindex, maybe_zindex)| {
+            (
+                root_id,
+                (
+                    maybe_global_zindex.map(|zindex| zindex.0).unwrap_or(0),
+                    maybe_zindex.map(|zindex| zindex.0).unwrap_or(0),
+                ),
+            )
+        })
+        .chain(global_nodes)
```

Comment:
Could you try doing this without chain? This combinator does come with a cost on a per-item basis.

Reply:
I was thinking that one collect() on a chained iterator would be more efficient than collecting and then extending.
Or do you mean that I could create the vec with:
```
let mut root_nodes = Vec::with_capacity(root_node_query.iter().len() + global_zindex_node_query.iter().len());
```
and then extend twice?

I'm not sure the difference would be measurable though except with an extremely artificial benchmark because most UI is only going to use one root node and I think even an extremely complex UI wouldn't need more than a dozen (but maybe oceantume can can suggest a use case, I'm not certain).

Reply:
The main thing I'd be concerned with is that it may break autoectorization or prevent collect from properly estimating the final output Vec size. Though you are right that it is very unlikely that we'll have too many roots.